Neo4j migration from 5.5.0 to 5.11.0 - Failed to invoke procedure `apoc.cypher.mapParallel2`: Caused by:RuntimeException: Error polling, timeout

I'm trying to migrate from Neo4j `5.5.0` to `5.11.0` (also tried `5.10.0`)

The following query works fine on Neo4j `5.5.0`:

    MATCH  ( childD:Vacancy )  WITH collect({`childD`: childD}) as childDDgRdgd  
    CALL apoc.cypher.mapParallel2(" WITH _.childD as childD 
    WHERE  ( coalesce(size(childD.employmentIds), 0) <= 0  
    OR  exists { MATCH (childD)-[:EMPLOYMENT_AS]-(req0:Requirable) 
    WHERE req0.id IN $reqParam0}  )  AND  ( coalesce(size(childD.locationIds), 0) <= 0  
    OR  exists { MATCH (childD)-[:LOCATED_IN]-(req1:Requirable) 
    WHERE req1.id IN $reqParam1}  )  AND  (childD.`status` = $status5)  
    AND  (childD.`active` = $active4)  AND ( (childD.`hourlyRateUsd` >= $hourlyRateUsd2)  
    OR  (childD.`salaryUsd` >= $salaryUsd3) ) 
    WITH childD RETURN childD  ", {`reqParam1`:[6, 7], `reqParam0`:[26], `active4`:true, `salaryUsd3`:13088, `hourlyRateUsd2`:96, `status5`:'APPROVED'}, childDDgRdgd, 6, 100) YIELD value as value   
    WITH value.childD as childD  
    return childD

But fails on Neo4j `5.11.0` with:

    Failed to invoke procedure `apoc.cypher.mapParallel2`: Caused by: java.lang.RuntimeException: Error polling, timeout of 10 seconds reached.

I also tried to extend the timeout to 100, but with the same result.

Has anything changed between Neo4j `5.5.0` and Neo4j `5.11.0`? Maybe some additional configurations are required or something like that?

For my tests, I'm running Neo4j in Docker with:                           

`<NEO4J_db_memory_transaction_total_max>0</NEO4J_db_memory_transaction_total_max>                               <NEO4J_server_memory_pagecache_size>4G</NEO4J_server_memory_pagecache_size>                                 <NEO4J_server_memory_heap_initial__size>5G</NEO4J_server_memory_heap_initial__size>                        <NEO4J_server_memory_heap_max__size>5G</NEO4J_server_memory_heap_max__size>
`

Please help, thanks!

**Steps to reproduce:**

Please start the following database with Neo4j 5.12.0 Community Edition. I run it in Docker.

[data.zip](https://github.com/neo4j/neo4j/files/12774362/data.zip)

database username: neo4j
database  password: neo4j12345

ans execute the following query:

`MATCH  ( childD:Profile )  WITH collect({`childD`: childD}) as childDDgRdgd  
CALL apoc.cypher.mapParallel2(" WITH _.childD as childD WHERE  (childD.`active` = $active0)  AND  (childD.`minInfoProvided` = $minInfoProvided1)  AND  (childD.`status` = $status2)  WITH childD RETURN childD  ", {`active0`:true, `status2`:'APPROVED', `minInfoProvided1`:true}, childDDgRdgd, 6, 10) YIELD value as value   
WITH value.childD as childD   
WITH childD  
WITH collect({`childD`:childD }) as aggregate WITH aggregate, size(aggregate) as count 
UNWIND aggregate as item 
WITH count, item.childD as childD  
ORDER BY childD.createdAt DESC SKIP 0 LIMIT 5 
OPTIONAL MATCH (childD)-[ru:CREATED_BY]->(u:User)  
RETURN count, childD AS decision`

You should receive:

`Failed to invoke procedure `apoc.cypher.mapParallel2`: Caused by: java.lang.RuntimeException: Error polling, timeout of 10 seconds reached.`

If not, reexecute the query until this error occurs.

Regarding:

`org.neo4j.driver.exceptions.DatabaseException: Access to record Node[1648,used=false,created=false,rel=-1,prop=-1,labels=Inline(0x0:[]),light,fixedRefs=false] went out of bounds of the page. The record size is 15 bytes, and the access was at offset 150 bytes into page 3, and the pages have a capacity of 8192 bytes. The mapped store file in question is /data/databases/neo4j/neostore.nodestore.db`

I don't know how to reliably reproduce this. it occasionally crashes during program execution.

--- a/extended/src/main/java/apoc/cypher/CypherExtended.java
+++ b/extended/src/main/java/apoc/cypher/CypherExtended.java
@@ -178,7 +178,7 @@ public class CypherExtended {
                 if (isPeriodicOperation(stmt)) {
                     Util.inThread(pools , () -> {
                         try {
-                            return db.executeTransactionally(stmt, params, result -> consumeResult(result, queue, addStatistics, timeout));
+                            return db.executeTransactionally(stmt, params, result -> consumeResult(result, queue, addStatistics, tx));
                         } catch (Exception e) {
                             collectError(queue, reportError, e, fileName);
                             return null;
@@ -188,7 +188,7 @@ public class CypherExtended {
                 else {
                     Util.inTx(db, pools, threadTx -> {
                         try (Result result = threadTx.execute(stmt, params)) {
-                            return consumeResult(result, queue, addStatistics, timeout);
+                            return consumeResult(result, queue, addStatistics, tx);
                         } catch (Exception e) {
                             collectError(queue, reportError, e, fileName);
                             return null;
@@ -231,7 +231,7 @@ public class CypherExtended {
             if (schemaOperation) {
                 Util.inTx(db, pools, txInThread -> {
                     try (Result result = txInThread.execute(stmt, params)) {
-                        return consumeResult(result, queue, addStatistics, timeout);
+                        return consumeResult(result, queue, addStatistics, tx);
                     } catch (Exception e) {
                         collectError(queue, reportError, e, fileName);
                         return null;
@@ -243,7 +243,7 @@ public class CypherExtended {
 
     private final static Pattern shellControl = Pattern.compile("^:?\\b(begin|commit|rollback)\\b", Pattern.CASE_INSENSITIVE);
 
-    private Object consumeResult(Result result, BlockingQueue<RowResult> queue, boolean addStatistics, long timeout) {
+    private Object consumeResult(Result result, BlockingQueue<RowResult> queue, boolean addStatistics, Transaction tx) {
         try {
             long time = System.currentTimeMillis();
             int row = 0;
@@ -372,14 +372,17 @@ public class CypherExtended {
     public Stream<MapResult> mapParallel2(@Name("fragment") String fragment, @Name("params") Map<String, Object> params, @Name("list") List<Object> data, @Name("partitions") long partitions,@Name(value = "timeout",defaultValue = "10") long timeout) {
         final String statement = withParamsAndIterator(fragment, params.keySet(), "_");
         tx.execute("EXPLAIN " + statement).close();
-        BlockingQueue<RowResult> queue = new ArrayBlockingQueue<>(100000);
+        int queueCapacity = 100000;
+        BlockingQueue<RowResult> queue = new ArrayBlockingQueue<>(queueCapacity);
+        ArrayBlockingQueue<Transaction> transactions = new ArrayBlockingQueue<>(queueCapacity);
         Stream<List<Object>> parallelPartitions = Util.partitionSubList(data, (int)(partitions <= 0 ? PARTITIONS : partitions), null);
         Util.inFuture(pools, () -> {
             long total = parallelPartitions
                 .map((List<Object> partition) -> {
-                    try (Transaction transaction = db.beginTx();
-                         Result result = transaction.execute(statement, parallelParams(params, "_", partition))) {
-                        return consumeResult(result, queue, false, timeout);
+                    Transaction transaction = db.beginTx();
+                    transactions.add(transaction);
+                    try (Result result = transaction.execute(statement, parallelParams(params, "_", partition))) {
+                        return consumeResult(result, queue, false, transaction);
                     } catch (Exception e) {
                         throw new RuntimeException(e);
                     }}
@@ -387,7 +390,10 @@ public class CypherExtended {
             queue.put(RowResult.TOMBSTONE);
             return total;
         });
-        return StreamSupport.stream(new QueueBasedSpliterator<>(queue, RowResult.TOMBSTONE, terminationGuard, (int)timeout),true).map((rowResult) -> new MapResult(rowResult.result));
+        
+        return StreamSupport.stream(new QueueBasedSpliterator<>(queue, RowResult.TOMBSTONE, terminationGuard, (int)timeout),true)
+                .map(rowResult -> new MapResult(rowResult.result))
+                .onClose(() -> transactions.forEach(Transaction::close));
     }
 
     public Map<String, Object> parallelParams(@Name("params") Map<String, Object> params, String key, List<Object> partition) {

--- a/extended/src/test/java/apoc/cypher/CypherEnterpriseExtendedTest.java
+++ b/extended/src/test/java/apoc/cypher/CypherEnterpriseExtendedTest.java
@@ -231,13 +231,23 @@ public class CypherEnterpriseExtendedTest {
 
     @Test
     public void testCypherMapParallel2WithResults() {
-        String query = """
-                MATCH (n:ReturnQuery) WITH COLLECT(n) AS list
-                CALL apoc.cypher.mapParallel2('MATCH (_)-[r:REL]->(o:Other) RETURN r, o', {}, list, 1)
-                YIELD value RETURN value""";
-        Map<String, Object> params = Map.of("file", SIMPLE_RETURN_QUERIES);
+        // this test could cause flaky errors, 
+        // so we need to run the query several times to make sure it always works
+        for (int i = 0; i < 30; i++) {
+            String query = """
+                    MATCH (n:ReturnQuery) WITH COLLECT(n) AS list
+                    CALL apoc.cypher.mapParallel2('MATCH (_)-[r:REL]->(o:Other) RETURN r, o', {}, list, 1)
+                    YIELD value RETURN value""";
+            Map<String, Object> params = Map.of("file", SIMPLE_RETURN_QUERIES);
+            testCypherMapParallelCommon(query, params);
 
-        testCypherMapParallelCommon(query, params);
+            session.writeTransaction(tx -> tx.run("MATCH (n) DETACH DELETE n"));
+        }
+
+        // Check that `SHOW TRANSACTIONS` just returns itself 
+        String showTransactionsQuery = "SHOW TRANSACTIONS";
+        testCall(session, showTransactionsQuery,
+                r -> assertEquals(showTransactionsQuery, r.get("currentQuery")));
     }
 
     public void testRunProcedureWithSimpleReturnResults(String query, Map<String, Object> params) {

--- a/extended/src/test/java/apoc/cypher/CypherExtendedTest.java
+++ b/extended/src/test/java/apoc/cypher/CypherExtendedTest.java
@@ -5,6 +5,7 @@ import apoc.util.TestUtil;
 import apoc.util.Util;
 import apoc.util.Utils;
 import apoc.util.collection.Iterators;
+import org.apache.commons.io.FileUtils;
 import org.junit.*;
 import org.junit.rules.ExpectedException;
 import org.neo4j.configuration.GraphDatabaseSettings;
@@ -19,6 +20,8 @@ import org.neo4j.test.rule.DbmsRule;
 import org.neo4j.test.rule.ImpermanentDbmsRule;
 
 import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
@@ -46,11 +49,11 @@ import static org.neo4j.driver.internal.util.Iterables.count;
  * @since 08.05.16
  */
 public class CypherExtendedTest {
-
+    public static final String IMPORT_DIR = "src/test/resources";
     @ClassRule
     public static DbmsRule db = new ImpermanentDbmsRule()
             .withSetting(GraphDatabaseSettings.allow_file_urls, true)
-            .withSetting(GraphDatabaseSettings.load_csv_file_url_root, new File("src/test/resources").toPath().toAbsolutePath());
+            .withSetting(GraphDatabaseSettings.load_csv_file_url_root, new File(IMPORT_DIR).toPath().toAbsolutePath());
 
     @Rule
     public ExpectedException thrown= ExpectedException.none();
@@ -226,7 +229,83 @@ public class CypherExtendedTest {
                     "RETURN value.title limit 5",
                 r -> assertEquals(5, Iterators.count(r)));
     }
-    
+
+    @Test
+    public void testIssue3751MapParallel2() {
+        int expected = 500;
+        db.executeTransactionally("UNWIND range(1, $int) as id CREATE (:Polling {id: id})",
+                Map.of("int", expected));
+
+        // the error is flaky, so we need to run the query several times to replicate it
+        for (int i = 0; i < 30; i++) {
+            testCallCount(db, """
+                            MATCH (n:Polling) WITH collect({childD: n}) as params \s
+                            CALL apoc.cypher.mapParallel2(" WITH _.childD as childD RETURN childD", {}, params, 6, 10)\s
+                            YIELD value RETURN value""",
+                    Map.of(),
+                    expected);
+        }
+
+        // Check that `SHOW TRANSACTIONS` just returns itself 
+        String showTransactionsQuery = "SHOW TRANSACTIONS";
+        testCall(db, showTransactionsQuery,
+                r -> assertEquals(showTransactionsQuery, r.get("currentQuery")));
+    }
+
+
+
+    @Test
+    public void testIssue3751RunFiles() {
+        int numEntities = 500;
+        db.executeTransactionally("UNWIND range(1, $int) as id CREATE (:Polling {id: id})",
+                Map.of("int", numEntities));
+
+        for (int i = 0; i < 30; i++) {
+            int numFiles = 30;
+            List<String> files = Collections.nCopies(numFiles, "parallel.cypher");
+
+            int expected = numEntities * numFiles;
+            testCallCount(db, "CALL apoc.cypher.runFiles($files, {statistics: false})",
+                    Map.of("files", files),
+                    expected);
+        }
+    }
+
+    @Test
+    public void testIssue3751RunSchemaFiles() throws IOException {
+        for (int i = 0; i < 15; i++) {
+            int numFiles = 10;
+            List<File> files = new ArrayList<>();
+
+            for (int fileIdx = 0; fileIdx < numFiles; fileIdx++) {
+                String id = i + "" + fileIdx;
+                File file = new File(IMPORT_DIR, "schema" + id);
+
+                String content = """
+                            CREATE INDEX index%1$s FOR (n:Person%1$s) ON (n.name%1$s);
+                            CREATE INDEX secondIndex%1$s FOR (n:Foo%1$s) ON (n.bar%1$s);
+                            CREATE INDEX thirdIndex%1$s FOR (n:Ajeje%1$s) ON (n.brazorf%1$s);
+                            """
+                        .formatted(id);
+                
+                FileUtils.writeStringToFile(file,
+                        content,
+                        StandardCharsets.UTF_8);
+                
+                files.add(file);
+            }
+
+            // 4 is the number of indexes
+            int expected = numFiles * 3;
+            List<String> fileNames = files.stream().map(File::getName).toList();
+            testCallCount(db, "CALL apoc.cypher.runSchemaFiles($files, {statistics: true})",
+                    Map.of("files", fileNames),
+                    expected);
+        
+            files.forEach(File::delete);
+        }
+        
+    }
     
     @Test
     public void testRunFileWithParameters() throws Exception {

--- a/extended/src/test/resources/parallel.cypher
+++ b/extended/src/test/resources/parallel.cypher
@@ -1,0 +1,1 @@
+MATCH (n:Polling) SET n.test = date() RETURN n;

